c10/core/ScalarType.h

```diff
@@ -217,6 +217,11 @@ static inline bool isIntegralType(ScalarType t) {
       t == ScalarType::Long || t == ScalarType::Short);
 }
 
+static inline bool isSignedType(ScalarType t) {
+  return (
+    t != ScalarType::Byte && t != ScalarType::Char && t != ScalarType::QUInt8);
```

Comment:
what isn't Char signed?

Reply:
also, we already have a an `is_signed` on Tensors, maybe you could just use that implementation rather than maintaining another list here (it uses std::numeric_limits).

Reply:
hm... I have a ScalarType here though not a tensor. Tensor also has `is_floating_point` but we have an `isFloatingType` below here too.

I thought a char was always an unsigned byte, but googling I see there are signed chars in C++...  Do we have signed character tensors?

Reply:
well `is_floating_point` just calls `isFloatingType` so that's more-or-less fine.

Here you actually have two different implementations.  And to answer your question about signed chars, look here: https://github.com/pytorch/pytorch/blob/30e03df63895600d37d77aa82e96b8b6db5842f2/aten/src/ATen/Dispatch.h#L205.

Reply:
ah interesting. I can refactor TypeProperties.cpp to have a function taking a ScalarType then.

This makes me think we may want additional restrictions on can_cast though. We explicitly decided we want to allow double->float regardless of losing information, but do we really want to allow, say, double->char?

Reply:
So I pointed this out in the `can_cast` call, but off the top of my head (haven't double checked) I thought we were going to do `promote` casting?  (i.e. same_kind || (safe-with-promote-semantics))

Reply:
sorry `double->char` wouldn't work because not same kind, I just meant larger drops in size, e.g. we allow long->int but are we sure we want to allow going all the way from long->char?

Reply:
numpy allows it.